File: frontend/lib/models/joseki.dart
import 'package:frontend/util/go_rule.dart';

const int boardSize = 19;
const int screenBoardSize = 14;

enum StoneColor { black, white, empty }

StoneColor reversedColor(StoneColor color) {
  if (color == StoneColor.black) {
    return StoneColor.white;
  } else if (color == StoneColor.white) {
    return StoneColor.black;
  } else {
    return StoneColor.empty;
  }
}

// pass:
class Stone {
  Stone(this.color, this.x, this.y, this.index) {
    assert(-1 <= x && x < boardSize);
    assert(-1 <= y && y < boardSize);
  }
  final StoneColor color;
  final int x;
  final int y;
  final int index;

  bool isPassed() {
    return x == -1 && y == -1;
  }

  Map<String, dynamic> toJson() {
    return {'color': color == StoneColor.black ? 0 : 1, 'x': x, 'y': y};
  }
}

typedef StoneList = List<Stone>;
typedef StoneMatrix = List<StoneList>;

class Joseki {
  StoneList stoneList = [];

  Joseki(StoneList stones) {
    stoneList = stones;
  }

  bool pushStone({required StoneColor color, required int x, required int y}) {
    Stone stone = Stone(color, x, y, stoneList.length);
    stoneList.add(stone);
    if (getProcessedBoard(stoneList).isNotEmpty) {
      return true;
    } else {
      stoneList.removeLast();
      return false;
    }
  }

  void clear() {
    stoneList.clear();
  }

  void popStone() {
    if (stoneList.isNotEmpty) {
      stoneList.removeLast();
    }
  }

  void popStones(int count) {
    for (int i = 0; i < count; i++) {
      popStone();
    }
  }

  List<Stone> getPassedStones() {
    return stoneList.where((stone) => stone.isPassed()).toList();
  }
}

class GobanState {
  Joseki joseki = Joseki([]);
  StoneMatrix stoneMatrix = getProcessedBoard([]);

  GobanState(Joseki newJoseki) {
    joseki = newJoseki;
    stoneMatrix = getProcessedBoard(joseki.stoneList);
  }
}

File: web/src/models/joseki.ts
import { getProcessedBoard } from "../utils/goGameRule";

export const boardSize = 19;
// export const screenBoardSize = 14;

export type StoneColor = "black" | "white" | "empty";

export const reversedColor = (color: StoneColor): StoneColor => {
  if (color == "black") {
    return "white";
  } else if (color == "white") {
    return "black";
  } else {
    return "empty";
  }
}

export class Stone {
  color: StoneColor;
  x: number;
  y: number;
  index: number;
  constructor(color: StoneColor, x: number, y: number, index: number) {
    if (!(0 <= x && x < boardSize && 0 <= y && y < boardSize)) {
      throw new Error("Stone position is out of board");
    }
    this.color = color;
    this.x = x;
    this.y = y;
    this.index = index;
  }

  // Map<String, dynamic> toJson() {
  //   return {'color': color == StoneColor.black ? 0 : 1, 'x': x, 'y': y};
  // }
}

export class Joseki {
  stoneList: Array<Stone> = [];

  constructor(stones: Array<Stone>) {
    this.stoneList = stones;
  }

  pushStone(color: StoneColor, x: number, y: number): boolean {
    const stone = new Stone(color, x, y, this.stoneList.length);
    this.stoneList.push(stone);
    if (getProcessedBoard(this.stoneList).length > 0) {
      return true;
    } else {
      this.stoneList.pop();
      return false;
    }
  }

  clear() {
    this.stoneList = [];
  }

  popStone() {
    if (this.stoneList.length > 0) {
      this.stoneList.pop();
    }
  }

  popStones(count: number) {
    for (let i = 0; i < count; i++) {
      this.popStone();
    }
  }
}

export class GobanState {
  joseki: Joseki = new Joseki([]);
  stoneMatrix: Array<Array<Stone>> = getProcessedBoard([]);

  constructor(newJoseki: Joseki) {
    this.joseki = newJoseki;
    this.stoneMatrix = getProcessedBoard(this.joseki.stoneList);
  }
}

export const generateEmptyMatrix = (): Array<Array<Stone>> => {
  const board: Array<Array<Stone>> = []
  for (let i = 0; i < boardSize; i++) {
    const row: Array<Stone> = []
    for (let j = 0; j < boardSize; j++) {
      row.push({ color: "empty", x: i, y: j, index: -1 })
    }
    board.push(row)
  }
  return board
}